File: py/1d_linear_convection.py
import numpy                       #here we load numpy
from matplotlib import pyplot      #here we load matplotlib
import time, sys                   #and load some utilities

nx = 41  # try changing this number from 41 to 81 and Run All ... what happens?
dx = 2 / (nx-1)
nt = 25    #nt is the number of timesteps we want to calculate
dt = .025  #dt is the amount of time each timestep covers (delta t)
c = 1      #assume wavespeed of c = 1

u = numpy.ones(nx)      #numpy function ones()
u[int(.5 / dx):int(1 / dx + 1)] = 2  #setting u = 2 between 0.5 and 1 as per our I.C.s

un = numpy.ones(nx) #initialize a temporary array
f = open("output.txt", "a")
for n in range(nt):  #loop for values of n from 0 to nt, so it will run nt times
    un = u.copy() ##copy the existing values of u into un
    for i in range(1, nx): ## you can try commenting this line and...
    # for i in range(nx): ## ... uncommenting this line and see what happens!
        u[i] = un[i] - c * dt / dx * (un[i] - un[i-1])
        f.write(str(u[i]) + '\n');
    pyplot.plot(numpy.linspace(0, 2, nx), u);
    pyplot.axis([0, 2, .5, 2.5])
    pyplot.pause(.0001)
    pyplot.cla()
pyplot.show()
f.close()

File: cpp/1d_linear_convection.cpp
#include <fstream>
#include <iostream>
#include <vector>
using namespace std;

int main() {
    double nx = 41;
    double dx = 2 / (nx - 1);
    int nt = 25;
    float dt = 0.025;
    int c = 1;
    double z = 1 / dx + 1;

    vector<double> u((int)nx);
    fill(u.begin(), u.end(), 1);
    fill(u.begin() + int(0.5 / dx), u.end() - int(1 / dx + 1), 2);

    vector<double> un((int)nx);

    ofstream f;
    f.open("output.txt");
    for (int n = 0; n < nt; n++) {
        un = u;
        for (int i = 1; i < nx; i++) {
            u[i] = un[i] - c * dt / dx * (un[i] - un[i - 1]);
            f << u.at(i) << endl;
        }
    }
    f.close();

    // for (int i = 0; i < u.size(); i++) {
    //     cout << u.at(i) << " ";
    // }
}